storage/dataflux/worksteal.go

```diff
@@ -130,9 +130,13 @@ func (w *worker) doWorkstealListing(ctx context.Context) error {
 				time.Sleep(sleepDurationWhenIdle)
 				continue
 			} else {
-				newRange := <-w.lister.ranges
-				<-w.idleChannel
-				w.updateWorker(newRange.startRange, newRange.endRange, active)
+				select {
+				case newRange := <-w.lister.ranges:
+					<-w.idleChannel
+					w.updateWorker(newRange.startRange, newRange.endRange, active)
+				case <-time.After(sleepDurationWhenIdle):
```

Comment:
This seems like kind of a stopgap?

Really I think this and the if case above at L129 should be a single for/select loop potentially. Could you try writing the code this way and let me know if it works?

I see why this was causing deadlocks now, because lister.ranges could be non-empty but then become empty before L133 is reached if other goroutines are waiting.

In general we should recv from a channel to check if it is ready, and use select to block on that or give an out if it is not. Checking and then receiving separately is non-atomic so seems reasonable that these kinds of issues will arise.

Reply:
Yes the code above can be fully deleted, just forgot to remove it in this push from my fork.